### PR TITLE
Elastic Horovod on Spark

### PR DESCRIPTION
**Pull request by EnricoMi**
This Elastic Horovod on Spark implementation is limited to fault-tolerant. It already has some code related to auto-scaling but this is incomplete for simplicity of first implementation.

Most notable change to existing code is the way how `SparkDriverService._ranks_to_indices` gets populated. Currently, `horovod.spark.runner` tries to replicate how MPI and Gloo will assign ranks to hosts and slots:

        # Determine the ranks to indices
        ranks_to_indices = []
        for host_hash in host_hashes:
            ranks_to_indices += driver.task_host_hash_indices()[host_hash]
        driver.set_ranks_to_indices(ranks_to_indices)

When MPI / Gloo diverge from this assumption, ranks and indices get mixed up. The new implementation detects the actual rank-index mapping on the task side where `host-hash`, `local-rank` and `rank` are known for MPI and Gloo. Given to the driver, the driver determines the `index` and populates the mapping. This is more robust than the current implementation. In elastic fault-tolerant or auto-scale mode, we cannot make these assumptions upfront. There, a `SparkRendevouzServer` intercepts the host allocation plan and provides the Spark driver with these information.

Note: contains https://github.com/horovod/horovod/pull/1861 so that PR should be merged into master and elastic before this. Once `elastic` branch is merged into `master`, this PR will be retargeted to `master`.

### CODE REVIEW

**Comment by EnricoMi** [comment]
@tgaddair why are we doing this in static Horovod on Spark:

        host_hashes = list(driver.task_host_hash_indices().keys())
        host_hashes.sort()
        while 0 not in driver.task_host_hash_indices()[host_hashes[0]]:
            host_hashes = host_hashes[1:] + host_hashes[:1]

Does the order of the host hashes matter? Is it to get task with `index == 0` to have `rank == 0`?

**Comment by EnricoMi** [comment]
reopen targeting horovod:master

File: horovod/spark/runner.py [hunk 1]
```diff
@@ -232,31 +283,130 @@ def run(fn, args=(), kwargs={}, num_proc=None, start_timeout=None,
...
+                                                    nics=nics,
+                                                    run_func_mode=True)
+
+    result_queue = queue.Queue(1)
+
+    # launch settings.num_proc / settings.max_np Spark tasks
+    spark_thread = _make_spark_thread(spark_context, spark_job_group, driver,
+                                      result_queue, settings, use_gloo=True, is_elastic=True)
+    try:
+        # TODO: why are we doing this in static and is it needed for elastic?
```
Comment (tgaddair) [inline horovod/spark/runner.py:381]: @alsrgv might be able to comment, as the original author of this code.

File: horovod/run/common/service/task_service.py [hunk 2]
```diff
@@ -168,7 +168,7 @@ def _handle(self, req, client_address):
             self._wait_cond.acquire()
             try:
                 while self._command_thread is None or self._command_thread.is_alive():
-                    self._wait_cond.wait(req.delay if req.delay >= 1.0 else 1.0)
+                    self._wait_cond.wait(req.delay if req.delay >= 0.1 else 0.1)
```
[the remark this answers is not in the document]
Comment (EnricoMi) [reply horovod/run/common/service/task_service.py:171]: introduced `WAIT_FOR_COMMAND_MIN_DELAY` and now using `max(req.delay, WAIT_FOR_COMMAND_MIN_DELAY)`

File: horovod/run/common/service/driver_service.py [hunk 3]
```diff
@@ -95,13 +107,25 @@ def _filter_by_ip(self, addresses, target_ip):
         return {}
 
     def all_task_addresses(self, index):
-        return self._all_task_addresses[index].copy()
+        self._wait_cond.acquire()
```
[the remark this answers is not in the document]
Comment (EnricoMi) [reply horovod/run/common/service/driver_service.py:110]: Have to use the `_wait_cond` here because we want to read these variables while they are not being modified in handling `RegisterTask*` requests. A different lock would not prevent that concurrency.

File: horovod/run/elastic/settings.py [hunk 4]
```diff
@@ -33,3 +33,10 @@ def __init__(self, discovery, min_np, max_np, elastic_timeout, **kwargs):
         self.min_np = min_np
         self.max_np = max_np
         self.elastic_timeout = elastic_timeout
+
+    # we do not serialize the discovery instance
+    # it is not needed on the worker and might not be serializable
+    def __getstate__(self):
```
[the remark this answers is not in the document]
Comment (EnricoMi) [reply horovod/run/elastic/settings.py:39]: Sounds good! #2022

File: horovod/spark/driver/driver_service.py [hunk 5]
```diff
@@ -68,24 +87,61 @@ def __init__(self, num_proc, fn, args, kwargs, key, nics):
...
+        if isinstance(req, SetLocalRankToRankRequest):
+            self._wait_cond.acquire()
+
+            try:
+                # get index for host and local_rank
+                indices = self._task_host_hash_indices[req.host]
+                index = indices[req.local_rank]
+
+                # remove earlier rank for this index
+                values = list(self._ranks_to_indices.values())
```
[the remark this answers is not in the document]
Comment (EnricoMi) [reply horovod/spark/driver/driver_service.py:108]: It returns a `ValuesView` which behaves like a `list` but it does not provide the `index` method, so turning it into a `list` first.

File: horovod/spark/driver/driver_service.py [hunk 6]
```diff
@@ -68,24 +87,61 @@ def __init__(self, num_proc, fn, args, kwargs, key, nics):
...
+                prev_pos = values.index(index) if index in values else None
+                if prev_pos is not None:
+                    prev_rank = list(self._ranks_to_indices.keys())[prev_pos]
```
[the remark this answers is not in the document]
Comment (EnricoMi) [reply horovod/spark/driver/driver_service.py:111]: Yes, this code relies on the fact that `keys()` and `values()` have the same order (insertion order is not needed here). Added a comment.

[the remark this answers is not in the document]
Comment (EnricoMi) [reply horovod/spark/driver/driver_service.py:111]: And `keys()` returns a `KeysView`, which does not provide the `[]`.

File: horovod/run/elastic/driver.py [hunk 7]
```diff
@@ -277,6 +277,7 @@ def run_worker():
         self._results.expect(thread)
 
     def _handle_worker_exit(self, slot_info, exit_code, timestamp):
+        # TODO: hosts that failed but are not reflected in host discovery yet are recorded here and make the job fail
```
[the remark this answers is not in the document]
Comment (EnricoMi) [reply horovod/run/elastic/driver.py:280]: I am not sure this is still an issue, I may have solved this somehow. Integration tests see this situation all the time and they work. Removed the comment.

File: horovod/spark/driver/driver_service.py [hunk 8]
```diff
@@ -28,6 +30,23 @@ def __init__(self, indices):
         """Task indices."""
 
 
+class SetLocalRankToRankRequest(object):
+    """Set local rank to tank."""
```
[the remark this answers is not in the document]
Comment (EnricoMi) [reply horovod/spark/driver/driver_service.py:34]: fixed

File: horovod/spark/driver/driver_service.py [hunk 9]
```diff
@@ -68,24 +87,61 @@ def __init__(self, num_proc, fn, args, kwargs, key, nics):
...
+        self._ranks_to_indices = {}
         self._spark_job_failed = False
 
     def _handle(self, req, client_address):
 
         if isinstance(req, TaskHostHashIndicesRequest):
             return TaskHostHashIndicesResponse(self._task_host_hash_indices[req.host_hash])
 
+        if isinstance(req, SetLocalRankToRankRequest):
+            self._wait_cond.acquire()
```
[the remark this answers is not in the document]
Comment (EnricoMi) [reply horovod/spark/driver/driver_service.py:100]: agree, simple `Lock` is sufficient here, fixed

File: horovod/spark/runner.py (hunk 1 above)
Comment (EnricoMi) [reply horovod/spark/runner.py:381]: I have removed those comments but would still like to know if this is something I need to do in elastic horovod. @alsrgv

File: horovod/torch/elastic.py [hunk 10]
```diff
@@ -44,6 +44,7 @@ def run(func):
 
 
 def _reset():
+    import os
```
Comment (tgaddair) [inline horovod/torch/elastic.py:47]: Why is this needed?

File: test/integration/data/elastic_tensorflow2_main.py [hunk 11]
```diff
@@ -32,11 +32,15 @@
...
 parser.add_argument('--epochs', type=int, default=3,
                     help='number of epochs')
+parser.add_argument('--epoch-wait', type=int, default=0,
+                    help='number of seconds each takes')
 parser.add_argument('--logfile', default='/tmp/logfile.txt',
                     help='log file to record results (one line per epoch)')
 parser.add_argument('--discovery-schedule', default='[]',
                     help='JSON string specifying schedule of host updates each epoch')
-parser.add_argument('--exit-schedule',
+parser.add_argument('--discovery-wait', type=int, default=3,
```
Comment (tgaddair) [inline test/integration/data/elastic_tensorflow2_main.py:41]: Nit: please specify units in the help message (e.g., seconds).

File: test/test_interactiverun.py [hunk 12]
```diff
@@ -59,6 +62,22 @@ def fn(a, b, c, d):
                                   "ret_val_of_rank_1",
                                   None], res2)
 
+    @pytest.mark.skipif(sys.version_info < (3, 0), reason='Elastic Horovod supported on Python3')
```
Comment (tgaddair) [inline test/test_interactiverun.py:65]: Shouldn't be needed anymore, since we dropped Python 2 support.

File: test/test_spark.py [hunk 13]
```diff
@@ -112,6 +319,27 @@ def fn():
                                         verbose=2)
                 self.assertListEqual([([0, 1], 0), ([0, 1], 1)], res)
 
+    """
+    Test that horovod.spark.run_elastic works properly in a simple setup.
+    """
+    @pytest.mark.skipif(sys.version_info < (3, 0), reason='Elastic Horovod supported on Python3')
```
Comment (tgaddair) [inline test/test_spark.py:325]: Same as above regarding Python 3 support.

File: horovod/torch/elastic.py (hunk 10 above)
Comment (EnricoMi) [reply horovod/torch/elastic.py:47]: debug leftover, removed

File: test/integration/data/elastic_tensorflow2_main.py (hunk 11 above)
Comment (EnricoMi) [reply test/integration/data/elastic_tensorflow2_main.py:41]: done: `Time in seconds the worker waits ...`

File: test/test_interactiverun.py (hunk 12 above)
Comment (EnricoMi) [reply test/test_interactiverun.py:65]: good catch, removed all python2 references, also one from `horovod.run.http.http_client.py`, though out of scope for this PR but while I am at it ...

File: test/test_spark.py (hunk 13 above)
Comment (EnricoMi) [reply test/test_spark.py:325]: done

**Comment by EnricoMi** [comment]
All comments addressed.
